server/core_purchase.go

```
@@ -0,0 +1,153 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"database/sql"
+	"nakama/pkg/iap"
+
+	"errors"
+
+	"encoding/json"
+
+	"github.com/satori/go.uuid"
+	"go.uber.org/zap"
+)
+
+type PurchaseService struct {
+	logger       *zap.Logger
+	db           *sql.DB
+	AppleClient  *iap.AppleClient
+	GoogleClient *iap.GoogleClient
+}
+
+func NewPurchaseService(jsonLogger *zap.Logger, multiLogger *zap.Logger, db *sql.DB, config *PurchaseConfig) *PurchaseService {
+	ac, err := iap.NewAppleClient(config.Apple.Password, config.Apple.Production, config.Apple.TimeoutMs)
+	if err != nil {
+		multiLogger.Warn("Skip initialising Apple in-app purchase provider", zap.Error(err))
+	} else {
+		if config.Apple.Production {
+			multiLogger.Info("Apple in-app purchase environment is set to Production priority.")
+		} else {
+			multiLogger.Info("Apple in-app purchase environment is set to Sandbox priority.")
+		}
+		multiLogger.Info("Successfully initiated Apple in-app purchase provider.")
+	}
+
+	gc, err := iap.NewGoogleClient(config.Google.PackageName, config.Google.ServiceKeyFilePath)
+	if err != nil {
+		multiLogger.Warn("Skip initialising Google in-app purchase provider", zap.Error(err))
+	}
+
+	return &PurchaseService{
+		logger:       jsonLogger,
+		db:           db,
+		AppleClient:  ac,
+		GoogleClient: gc,
+	}
+}
+
+func (p *PurchaseService) ValidateApplePurchase(userID uuid.UUID, purchase *iap.ApplePurchase) *iap.PurchaseVerifyResponse {
+	r, appleReceipt := p.AppleClient.Verify(purchase)
+	if !r.Success {
+		return r
+	}
+
+	//TODO: Improvement - Process more than one in-app receipts
+	inAppReceipt := appleReceipt.InApp[0]
+	p.checkUser(userID, r, 1, inAppReceipt.TransactionID)
+
+	if r.Success && !r.SeenBefore {
+		err := p.savePurchase(userID.Bytes(), 1, inAppReceipt.ProductID, inAppReceipt.TransactionID, purchase.ReceiptData, r.Data)
```

@zyro I believe multiple writes over the same subset of the keyspace in cockroach will be coordinated via the same leaserholder (and therefore serialized as writes). I'll have to double check though.

@mofirouz Is there a way we could return an error which indicated the database server could not be reached with IAP validation and let the game client decide how to handle it?